### PR TITLE
Split up core/test/iter.rs into multiple files

### PR DESCRIPTION
**Pull request by danii**
This PR removes the `// ignore-tidy-filelength` at the top of [iter.rs](https://github.com/rust-lang/rust/blob/04f44fb9232dc960d213d0df4a203c387215a5ff/library/core/tests/iter.rs) by splitting it into several sub files. I have split the file per test, based on what I felt the test was really trying to test.
Addresses issue #60302.
- [associated_util.rs](https://github.com/rust-lang/rust/blob/d29180a8ed2752c5dfa2bca1d28ca05dc9bf6bf1/library/core/tests/iter/associated_util.rs) - For testing the module level functions. (Maybe should be renamed?)
- [collection.rs](https://github.com/rust-lang/rust/blob/d29180a8ed2752c5dfa2bca1d28ca05dc9bf6bf1/library/core/tests/iter/collection.rs) - For testing methods that use the values of all the items in an iterator, and creates one value from them, whether it be a Vec of all the values, or a number that represents the sum.
- [mod.rs](https://github.com/rust-lang/rust/blob/d29180a8ed2752c5dfa2bca1d28ca05dc9bf6bf1/library/core/tests/iter/mod.rs) - For utility structs used in all tests, and other tests I didn't know where to place.
- [range.rs](https://github.com/rust-lang/rust/blob/d29180a8ed2752c5dfa2bca1d28ca05dc9bf6bf1/library/core/tests/iter/range.rs) - For testing ranges.
- [transformation.rs](https://github.com/rust-lang/rust/blob/d29180a8ed2752c5dfa2bca1d28ca05dc9bf6bf1/library/core/tests/iter/transformation.rs) - For testing methods that transform all the values in an iterator.
- [util.rs](https://github.com/rust-lang/rust/blob/d29180a8ed2752c5dfa2bca1d28ca05dc9bf6bf1/library/core/tests/iter/util.rs) - For testing methods that provide utility in more specific use cases.

"Programatically"
-----------------------
You may have noticed I "Programatically" split up the file. [This is how](https://gist.github.com/danii/a58b3bcafa9faf1c3e4b01ad7495baf4) I managed to do that. 😛 Hope that's fine.

### CODE REVIEW

**Comment by cuviper** [comment]
I'm fine with the idea of splitting this up, but the current subdivisions don't make much sense to me. The rough principal I would use is how easy someone could guess where to find related test cases. Like, suppose I just fixed something in `Chain`, then is it obvious where I would find tests focused on chain? (Hopefully they're together!)

(At the end of the day, this will all be very subjective, but it's also _just_ tests -- not an API we have to commit to.)

> * [associated_util.rs](https://github.com/rust-lang/rust/blob/d29180a8ed2752c5dfa2bca1d28ca05dc9bf6bf1/library/core/tests/iter/associated_util.rs) - For testing the module level functions. (Maybe should be renamed?)

This group seems fine, but I'd probably just call it `functions.rs`.

> * [collection.rs](https://github.com/rust-lang/rust/blob/d29180a8ed2752c5dfa2bca1d28ca05dc9bf6bf1/library/core/tests/iter/collection.rs) - For testing methods that use the values of all the items in an iterator, and creates one value from them, whether it be a Vec of all the values, or a number that represents the sum.

There's one `test_collect` that's on the nose, but the rest of these don't make sense to me. There are tests of folds, sums, products, min/max -- I would not have thought to look in `collection.rs` for any of these.

> * [mod.rs](https://github.com/rust-lang/rust/blob/d29180a8ed2752c5dfa2bca1d28ca05dc9bf6bf1/library/core/tests/iter/mod.rs) - For utility structs used in all tests, and other tests I didn't know where to place.

That's fine for `mod.rs` to be the default catch-all for tests, but I wouldn't bother with the `mod other` nesting. As for the utility structs, I think they only belong here if they would be shared by multiple different test modules, otherwise keep them close to their tests so it's clear what those are doing.

> * [range.rs](https://github.com/rust-lang/rust/blob/d29180a8ed2752c5dfa2bca1d28ca05dc9bf6bf1/library/core/tests/iter/range.rs) - For testing ranges.

This one is good, they're all range iterators.

> * [transformation.rs](https://github.com/rust-lang/rust/blob/d29180a8ed2752c5dfa2bca1d28ca05dc9bf6bf1/library/core/tests/iter/transformation.rs) - For testing methods that transform all the values in an iterator.

This is ambiguous to me -- I wouldn't guess what to find in there, and it looks like quite a mix.

> * [util.rs](https://github.com/rust-lang/rust/blob/d29180a8ed2752c5dfa2bca1d28ca05dc9bf6bf1/library/core/tests/iter/util.rs) - For testing methods that provide utility in more specific use cases.

This too, I find unintuitive.

I guess the point is that I don't really care what a test is _doing_, since they are usually invented scenarios that aren't the point. What matters is to group them by **what** they are testing. If I modify `Chain` or `Zip`, I want to find all the tests centered around that particular iterator. Stuff like `min`/`max` and maybe `eq`/`lt`/etc. can go together, but I would make it obvious like `ordering.rs`.

I think it's fine if these end up even smaller, rather than trying to define very broad groups. If it's _too_ small (subjective), just leave it at the top level.

**Comment by cuviper** [comment]
You could just mirror the layout of `core/src/iter/`, so the subjective organization is already decided for you, assuming you can figure out the focus of each particular test.

**Comment by Daniihh** [comment]
I've taken your suggestion to mirror the layout of `core/src/iter` to heart, and have decided to resort everything in `fns.txt` to programatically split it again.
Do you mind reviewing my [fns.txt](https://gist.github.com/Daniihh/a58b3bcafa9faf1c3e4b01ad7495baf4) in case there's any other big changes you want me to make?
Format of `fns.txt` is simple, `<resulting file> fn <function name in iter.rs>` for each line.
Yes, those file names do include `::`, but I'll move them manually into the right folders and rename them afterwards.

I have tried to mirror it as close as I can, but it did get fuzzy. I decided to put tests relating to a trait or an adapter in the module corresponding to it. I also decided to add new modules for some things if they ended up in the modules `core::iter::traits` and `core::iter::adapters` (excluding submodules). If you'd like, I can remove those extra submodules.

I feel this is definitely a lot better from my first attempt, because they aren't split into broad groups anymore, and it did feel like that was a bit of an issue. :)

**Comment by cuviper** [comment]
> Do you mind reviewing my [fns.txt](https://gist.github.com/Daniihh/a58b3bcafa9faf1c3e4b01ad7495baf4) in case there's any other big changes you want me to make?

This seems to be gone now, 404.

**Comment by danii** [comment]
Oh oops, I recently changed my name...
https://gist.github.com/danii/a58b3bcafa9faf1c3e4b01ad7495baf4
Surprised that doesn't temporarily redirect like repository links.

**Comment by cuviper** [comment]
@danii That breakdown looks good to me!

**Comment by crlf0710** [comment]
@danii Ping from triage, this seems need a rebase to continue. Would you mind doing so? Thanks!

**Comment by danii** [comment]
@rustbot label -S-waiting-on-author +S-waiting-on-review

After a season, this should be ready to rebase, I hope.

File: library/core/tests/iter.rs [hunk 1]
```diff
@@ -1,3681 +0,0 @@
...
-    let mut i = 0;
-    for x in it {
-        assert_eq!(x, ys[i]);
-        i += 1;
-    }
-    assert_eq!(i, ys.len());
-}
-
-/// Tests `FlatMap::fold` with items already picked off the front and back,
-/// to make sure all parts of the `FlatMap` are folded correctly.
```
Comment (cuviper) [inline library/core/tests/iter.rs:1543]: This doc-comment got lost in the move.

File: library/core/tests/iter.rs [hunk 2]
```diff
@@ -1,3681 +0,0 @@
...
-    let mut i = 0;
-    for x in it {
-        assert_eq!(x, ys[i]);
-        i += 1;
-    }
-    assert_eq!(i, ys.len());
-}
-
-/// Tests `Flatten::fold` with items already picked off the front and back,
-/// to make sure all parts of the `Flatten` are folded correctly.
```
Comment (cuviper) [inline library/core/tests/iter.rs:1581]: This doc-comment got lost in the move.

File: library/core/tests/iter/adapters/map.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
+use core::iter::*;
+
+#[test]
+fn test_find_map() {
```
Comment (cuviper) [inline library/core/tests/iter/adapters/map.rs:4]: This is not related to `map` or `Map` -- I think it belongs in the main `traits/iterator.rs`.

File: library/core/tests/iter/adapters/map.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    assert_eq!((0..10).map(|x| x + 3).try_rfold(7, f), (3..13).try_rfold(7, f));
+
+    let mut iter = (0..40).map(|x| x + 10);
+    assert_eq!(iter.try_fold(0, i8::checked_add), None);
+    assert_eq!(iter.next(), Some(20));
+    assert_eq!(iter.try_rfold(0, i8::checked_add), None);
+    assert_eq!(iter.next_back(), Some(46));
+}
+#[test]
+fn test_filter_map_try_folds() {
```
Comment (cuviper) [inline library/core/tests/iter/adapters/map.rs:38]: This belongs in `filter_map.rs`.

File: library/core/tests/iter/adapters/step.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
+use core::iter::*;
```
Comment (cuviper) [inline library/core/tests/iter/adapters/step.rs:1]: `Step` is a trait, so I guess this file should move to `traits/`.

File: library/core/tests/iter/mod.rs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    {
+        let iter = (0..5).map(|_| {
+            x += 1;
+        });
+        ().extend(iter);
+    }
+    assert_eq!(x, 5);
+}
+#[test]
+fn test_try_fold_specialization_intersperse_err() {
```
Comment (cuviper) [inline library/core/tests/iter/mod.rs:81]: This should go in `intersperse.rs`.

File: library/core/tests/iter/traits/double_ended.rs [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+#[test]
+fn test_iterator_rev_nth() {
+    let v: &[_] = &[0, 1, 2, 3, 4];
+    for i in 0..v.len() {
+        assert_eq!(v.iter().rev().nth(i).unwrap(), &v[v.len() - 1 - i]);
+    }
+    assert_eq!(v.iter().rev().nth(v.len()), None);
+}
+#[test]
+fn test_iterator_len() {
```
Comment (cuviper) [inline library/core/tests/iter/traits/double_ended.rs:18]: This test has nothing to do with being double-ended.

File: library/core/tests/iter/traits/double_ended.rs [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+#[test]
+fn test_rev() {
+    let xs = [2, 4, 6, 8, 10, 12, 14, 16];
+    let mut it = xs.iter();
+    it.next();
+    it.next();
+    assert!(it.rev().cloned().collect::<Vec<isize>>() == vec![16, 14, 12, 10, 8, 6]);
+}
+#[test]
+fn test_double_ended_map() {
```
Comment (cuviper) [inline library/core/tests/iter/traits/double_ended.rs:33]: The `rev`-related tests seem fine to keep in this file, but I think the rest are better associated with their specific type to test their implementation -- e.g. this one should be in `adapters/map.rs`.

File: library/core/tests/iter/traits/marker.rs [hunk 9]
```diff
@@ -0,0 +1,17 @@
+#[test]
+fn test_fuse() {
```
Comment (cuviper) [inline library/core/tests/iter/traits/marker.rs:2]: Why not in `fuse.rs`?

File: library/core/tests/iter/traits/mod.rs [hunk 10]
```diff
@@ -0,0 +1,39 @@
+mod accum;
+mod collect;
+mod double_ended;
+mod iterator;
+mod marker;
+
+#[test]
+fn test_rposition() {
```
Comment (cuviper) [inline library/core/tests/iter/traits/mod.rs:8]: These are all appropriate for `double_ended.rs`.

**Review by cuviper — CHANGES_REQUESTED** [review]
It's difficult to be sure that nothing is lost, but I'm concerned that the net line count is reduced:

    33 files changed, 3574 insertions(+), 3681 deletions(-)

If anything, I'd expect it to increase for the repetition of `use` statements in each new file. At least some of the loss will be simple whitespace between functions though, which I would like to see fixed.

File: library/core/tests/iter/traits/marker.rs (hunk 9 above)
Comment (danii) [reply library/core/tests/iter/traits/marker.rs:2]: My bad, Lol, I think I might have wrote marker in the input file for my splitter program to remind myself to come back to it, but I lost it amidst the chaos.

**Review by cuviper — APPROVED** [review]
Looks great, thanks! The added documentation is very welcome too.

**Comment by danii** [comment]
Ay, nice! I know I sat on this pull request for too long, but now I'm feeling great. Thank you.
